File: protocol.py
"""
消息协议定义
"""
import json
import time
from enum import Enum
from typing import Any, Dict, List, Optional, Union
from pydantic import BaseModel, Field


class MessageType(str, Enum):
    """消息类型枚举"""
    # 注册与认证
    AGENT_REGISTER = "agent.register"
    AGENT_REGISTER_ACK = "agent.register_ack"

    # 心跳
    HEARTBEAT = "heartbeat"
    HEARTBEAT_ACK = "heartbeat.ack"

    # 任务相关
    TASK_EXECUTE = "task.execute"
    TASK_STARTED = "task.started"
    TASK_PROGRESS = "task.progress"
    TASK_COMPLETED = "task.completed"
    TASK_FAILED = "task.failed"
    TASK_CANCEL = "task.cancel"
    TASK_CANCELLED = "task.cancelled"

    # 错误
    ERROR = "error"

    # 用户确认
    USER_CONFIRMATION_REQUEST = "user_confirmation.request"
    USER_CONFIRMATION_RESPONSE = "user_confirmation.response"


class ConfirmationOption(BaseModel):
    """确认选项"""
    label: str
    value: str


class UserConfirmationRequest(BaseModel):
    """用户确认请求"""
    request_id: str
    task_id: str
    title: str
    message: str
    prompt: str
    options: list[ConfirmationOption] = []
    timeout: int = 300  # 超时时间（秒）


class UserConfirmationResponse(BaseModel):
    """用户确认回应"""
    request_id: str
    task_id: str
    value: str
    timestamp: float = Field(default_factory=time.time)


class TaskOptions(BaseModel):
    """任务选项"""
    model: str = "sonnet"
    max_turns: int = 10
    effort: str = ""
    allowed_tools: Optional[List[str]] = None
    output_format: str = "text"
    timeout: int = 300
    continue_last: bool = False
    session_id: Optional[str] = None


class TaskPayload(BaseModel):
    """任务执行负载"""
    prompt: str
    context: Optional[str] = None
    workdir: str = "."
    options: TaskOptions = Field(default_factory=TaskOptions)


class TaskResult(BaseModel):
    """任务结果"""
    success: bool
    result: str = ""
    structured_output: Optional[Dict[str, Any]] = None
    usage: Dict[str, Any] = Field(default_factory=dict)
    duration_ms: int = 0
    num_turns: int = 0
    session_id: Optional[str] = None


class TaskProgress(BaseModel):
    """任务进度"""
    turn: int = 0
    max_turns: int = 10
    status: str = "thinking"
    message: Optional[str] = None


class Message(BaseModel):
    """基础消息"""
    type: MessageType
    id: Optional[str] = None
    payload: Dict[str, Any] = Field(default_factory=dict)
    timestamp: float = Field(default_factory=time.time)

    def to_json(self) -> str:
        """转换为JSON字符串"""
        return self.model_dump_json(by_alias=True)

    @classmethod
    def from_json(cls, json_str: str) -> "Message":
        """从JSON字符串解析"""
        data = json.loads(json_str)
        return cls(**data)


def build_register_message(client_id: str, version: str,
                           claude_version: str,
                           supported_tools: List[str]) -> Message:
    """构建注册消息"""
    return Message(
        type=MessageType.AGENT_REGISTER,
        payload={
            "client_id": client_id,
            "version": version,
            "capabilities": {
                "claude_version": claude_version,
                "supported_tools": supported_tools
            }
        }
    )


def build_heartbeat_message(status: str = "idle",
                            active_tasks: int = 0) -> Message:
    """构建心跳消息"""
    return Message(
        type=MessageType.HEARTBEAT,
        payload={
            "status": status,
            "active_tasks": active_tasks
        }
    )


def build_task_started_message(task_id: str) -> Message:
    """构建任务开始消息"""
    return Message(
        type=MessageType.TASK_STARTED,
        id=task_id,
        payload={
            "started_at": time.time()
        }
    )


def build_task_progress_message(task_id: str,
                                progress: TaskProgress) -> Message:
    """构建任务进度消息"""
    return Message(
        type=MessageType.TASK_PROGRESS,
        id=task_id,
        payload=progress.model_dump()
    )


def build_task_completed_message(task_id: str,
                                 result: TaskResult) -> Message:
    """构建任务完成消息"""
    return Message(
        type=MessageType.TASK_COMPLETED,
        id=task_id,
        payload=result.model_dump()
    )


def build_task_failed_message(task_id: str,
                              error: str,
                              error_code: str = "UNKNOWN",
                              partial_output: str = "") -> Message:
    """构建任务失败消息"""
    return Message(
        type=MessageType.TASK_FAILED,
        id=task_id,
        payload={
            "error": error,
            "error_code": error_code,
            "partial_output": partial_output
        }
    )


def build_task_cancelled_message(task_id: str) -> Message:
    """构建任务取消消息"""
    return Message(
        type=MessageType.TASK_CANCELLED,
        id=task_id,
        payload={
            "cancelled_at": time.time()
        }
    )


def build_error_message(error: str,
                        error_code: str = "UNKNOWN") -> Message:
    """构建错误消息"""
    return Message(
        type=MessageType.ERROR,
        payload={
            "error": error,
            "error_code": error_code
        }
    )


def build_user_confirmation_request(request: UserConfirmationRequest) -> Message:
    """构建用户确认请求消息"""
    return Message(
        type=MessageType.USER_CONFIRMATION_REQUEST,
        id=request.task_id,
        payload=request.model_dump()
    )


def build_user_confirmation_response(response: UserConfirmationResponse) -> Message:
    """构建用户确认回应消息"""
    return Message(
        type=MessageType.USER_CONFIRMATION_RESPONSE,
        id=response.task_id,
        payload=response.model_dump()
    )

File: node-client/src/protocol.ts
/**
 * 消息协议定义 — 与 Python 版 protocol.py 完全兼容
 */

// ---------------------------------------------------------------------------
// 消息类型
// ---------------------------------------------------------------------------

export type MessageType =
  | 'agent.register'
  | 'agent.register_ack'
  | 'heartbeat'
  | 'heartbeat.ack'
  | 'task.execute'
  | 'task.started'
  | 'task.progress'
  | 'task.event'
  | 'task.completed'
  | 'task.failed'
  | 'task.cancel'
  | 'task.cancelled'
  | 'error'
  | 'user_confirmation.request'
  | 'user_confirmation.response';

// ---------------------------------------------------------------------------
// 数据模型接口
// ---------------------------------------------------------------------------

export interface ConfirmationOption {
  label: string;
  value: string;
}

export interface UserConfirmationRequest {
  request_id: string;
  task_id: string;
  title: string;
  message: string;
  prompt: string;
  options: ConfirmationOption[];
  timeout: number;
  source?: string;
  tool_name?: string;
  tool_input?: Record<string, unknown>;
  tool_use_id?: string;
}

export interface UserConfirmationResponse {
  request_id: string;
  task_id: string;
  value: string;
  timestamp: number;
}

export interface TaskOptions {
  model: string;
  max_turns: number;
  effort: string | null;
  allowed_tools: string[] | null;
  output_format: string;
  timeout: number;
  continue_last: boolean;
  session_id: string | null;
  mode: string;
}

export interface TaskPayload {
  prompt: string;
  context?: string | null;
  workdir: string;
  options: TaskOptions;
}

export interface TaskResult {
  success: boolean;
  result: string;
  structured_output: Record<string, unknown> | null;
  usage: Record<string, unknown>;
  duration_ms: number;
  num_turns: number;
  session_id: string | null;
}

export interface TaskProgress {
  turn: number;
  max_turns: number;
  status: string;
  message: string | null;
}

export interface TaskEvent {
  task_id: string;
  seq: number;
  event_type: string;
  payload: Record<string, unknown>;
  timestamp: number;
}

export interface Message {
  type: MessageType;
  id: string | null;
  payload: Record<string, unknown>;
  timestamp: number;
}

// ---------------------------------------------------------------------------
// 序列化 / 反序列化
// ---------------------------------------------------------------------------

export function messageToJson(msg: Message): string {
  return JSON.stringify(msg);
}

export function messageFromJson(jsonStr: string): Message {
  return JSON.parse(jsonStr) as Message;
}

// ---------------------------------------------------------------------------
// 时间戳辅助
// ---------------------------------------------------------------------------

function unixTimestamp(): number {
  return Date.now() / 1000;
}

// ---------------------------------------------------------------------------
// 构建函数
// ---------------------------------------------------------------------------

export function buildRegisterMessage(
  clientId: string,
  version: string,
  claudeVersion: string,
  supportedTools: string[],
): Message {
  return {
    type: 'agent.register',
    id: null,
    payload: {
      client_id: clientId,
      version,
      capabilities: {
        claude_version: claudeVersion,
        supported_tools: supportedTools,
      },
    },
    timestamp: unixTimestamp(),
  };
}

export function buildHeartbeatMessage(
  status: string = 'idle',
  activeTasks: number = 0,
): Message {
  return {
    type: 'heartbeat',
    id: null,
    payload: { status, active_tasks: activeTasks },
    timestamp: unixTimestamp(),
  };
}

export function buildTaskStartedMessage(taskId: string): Message {
  return {
    type: 'task.started',
    id: taskId,
    payload: { started_at: unixTimestamp() },
    timestamp: unixTimestamp(),
  };
}

export function buildTaskProgressMessage(
  taskId: string,
  progress: TaskProgress,
): Message {
  return {
    type: 'task.progress',
    id: taskId,
    payload: { ...progress },
    timestamp: unixTimestamp(),
  };
}

export function buildTaskEventMessage(
  taskId: string,
  seq: number,
  eventType: string,
  payload: Record<string, unknown>,
): Message {
  return {
    type: 'task.event',
    id: taskId,
    payload: {
      task_id: taskId,
      seq,
      event_type: eventType,
      payload,
      timestamp: unixTimestamp(),
    },
    timestamp: unixTimestamp(),
  };
}

export function buildTaskCompletedMessage(
  taskId: string,
  result: TaskResult,
): Message {
  return {
    type: 'task.completed',
    id: taskId,
    payload: { ...result },
    timestamp: unixTimestamp(),
  };
}

export function buildTaskFailedMessage(
  taskId: string,
  error: string,
  errorCode: string = 'UNKNOWN',
  partialOutput: string = '',
): Message {
  return {
    type: 'task.failed',
    id: taskId,
    payload: { error, error_code: errorCode, partial_output: partialOutput },
    timestamp: unixTimestamp(),
  };
}

export function buildTaskCancelledMessage(taskId: string): Message {
  return {
    type: 'task.cancelled',
    id: taskId,
    payload: { cancelled_at: unixTimestamp() },
    timestamp: unixTimestamp(),
  };
}

export function buildErrorMessage(
  error: string,
  errorCode: string = 'UNKNOWN',
): Message {
  return {
    type: 'error',
    id: null,
    payload: { error, error_code: errorCode },
    timestamp: unixTimestamp(),
  };
}

export function buildUserConfirmationRequest(
  request: UserConfirmationRequest,
): Message {
  return {
    type: 'user_confirmation.request',
    id: request.task_id,
    payload: { ...request },
    timestamp: unixTimestamp(),
  };
}

export function buildUserConfirmationResponse(
  response: UserConfirmationResponse,
): Message {
  return {
    type: 'user_confirmation.response',
    id: response.task_id,
    payload: { ...response },
    timestamp: unixTimestamp(),
  };
}
